### PR TITLE
smoke test for CSI driver

### PR DESCRIPTION
**Pull request by jitendar-singh**
- BDD test framework based on behave for python
- Define the scenarios as feature file and add step definitions to test the features
- openshift.py - This module is used as a wrapper for openshift cli,one can create the Openshift object & use the functionalities to check the resources.
-  various other modules like project.py and command.py for added functionalities
- README.md describes the framework
- test environment setup script added in hack/smoke-test.sh
- adding smoke as target in Makefile so that smoke test can be triggered using make smoke command

### CODE REVIEW

File: Makefile
Comment (coreydaley) [inline Makefile:91]: +1, I didn't know about putting the `@` in front of the command!

File: hack/smoke-test.sh
Comment (coreydaley) [inline hack/smoke-test.sh:35]:
```suggestion
echo "Logs directory created at "${LOGS_DIR}/smoke
```

File: smoke/README.md
Comment (gabemontero) [inline smoke/README.md:5]:
```suggestion
We are using this framework to achieve the QA verification of any Jira Stories development takes on for the OpenShift Shared Resource CSI Driver, as well as any regression test suites the team wants to cultivate for the OpenShift Shared Resource CSI Driver.
```

File: hack/smoke-test.sh
Comment (gabemontero) [inline hack/smoke-test.sh:7]:
```suggestion
export OUTPUT_DIR=$(pwd)/_output
```

Comment (gabemontero) [reply hack/smoke-test.sh:7]: go with _output so `make  clean` cleans this up as well

File: smoke/README.md
Comment (gabemontero) [inline smoke/README.md:9]: You need to supply a workflow on *HOW* a new smoke test is added 
- add the feature file in a certain dir
- add the py file in a certain dir
- any naming conventions
- any refs to `behave` docs that might help new writers 
- refs and brief explanations for all the nice python utilities you've added for interacting with openshift like openshift.py., projects.py, steps.py, etc.

Similar to what @otaviof did for the BATS based e2e framework for shipwright CLI, like at https://github.com/shipwright-io/cli/blob/main/docs/testing.md#bats and https://github.com/shipwright-io/cli/blob/main/docs/testing.md#test-cases

Comment (gabemontero) [reply smoke/README.md:9]: And make sure you also cover all of my questions from https://github.com/openshift/csi-driver-shared-resource/pull/62#discussion_r758516919 as well wrt describing the process and methodology here moving forward

**Review by gabemontero — CHANGES_REQUESTED** [review]
ok @jitendar-singh here is my initial list of additions I'd like for you to make 

even though I've asked for more, please understand I'm liking how the dev/QA contract wrt the old, only "doc test cases in polarion model" we used to have will be improved upon here.

as we've discussed in team meetings, there will still be a place for polarion, but I'm liking the value add here and you building on what you've been doing with Jenkins

File: smoke/README.md
Comment (jitendar-singh) [reply smoke/README.md:8]: I have addressed the concerns and updated the README file

File: hack/smoke-test.sh
Comment (gabemontero) [inline hack/smoke-test.sh:7]: running `shellcheck` on your script here produced these results (install `shellcheck` if you do not have it and address theses): 

```
$ shellcheck smoke-test.sh 

In smoke-test.sh line 7:
export OUTPUT_DIR=$(pwd)/_output
       ^--------^ SC2155: Declare and assign separately to avoid masking return values.


In smoke-test.sh line 15:
python3 -m venv ${PYTHON_VENV_DIR}
                ^----------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
python3 -m venv "${PYTHON_VENV_DIR}"


In smoke-test.sh line 16:
${PYTHON_VENV_DIR}/bin/pip install --upgrade setuptools
^----------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
"${PYTHON_VENV_DIR}"/bin/pip install --upgrade setuptools


In smoke-test.sh line 17:
${PYTHON_VENV_DIR}/bin/pip install --upgrade pip
^----------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
"${PYTHON_VENV_DIR}"/bin/pip install --upgrade pip


In smoke-test.sh line 19:
echo "test-namespace-$(uuidgen | tr '[:upper:]' '[:lower:]' | head -c 8)" > ${OUTPUT_DIR}/test-namespace
                                                                            ^-----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
echo "test-namespace-$(uuidgen | tr '[:upper:]' '[:lower:]' | head -c 8)" > "${OUTPUT_DIR}"/test-namespace


In smoke-test.sh line 20:
if [[ $OPENSHIFT_CI == true ]]; then
   ^-------------------------^ SC3010: In POSIX sh, [[ ]] is undefined.


In smoke-test.sh line 22:
    oc project -q > ${OUTPUT_DIR}/test-namespace
                    ^-----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
    oc project -q > "${OUTPUT_DIR}"/test-namespace


In smoke-test.sh line 24:
export TEST_NAMESPACE=$(cat ${OUTPUT_DIR}/test-namespace)
       ^------------^ SC2155: Declare and assign separately to avoid masking return values.
                            ^-----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
export TEST_NAMESPACE=$(cat "${OUTPUT_DIR}"/test-namespace)


In smoke-test.sh line 28:
kubectl delete namespace ${TEST_NAMESPACE} --timeout=45s --wait
                         ^---------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
kubectl delete namespace "${TEST_NAMESPACE}" --timeout=45s --wait


In smoke-test.sh line 29:
kubectl create namespace ${TEST_NAMESPACE}
                         ^---------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
kubectl create namespace "${TEST_NAMESPACE}"


In smoke-test.sh line 31:
mkdir -p ${LOGS_DIR}/smoke-tests-logs
         ^---------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
mkdir -p "${LOGS_DIR}"/smoke-tests-logs


In smoke-test.sh line 32:
mkdir -p ${OUTPUT_DIR}/smoke-tests-output
         ^-----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
mkdir -p "${OUTPUT_DIR}"/smoke-tests-output


In smoke-test.sh line 33:
touch ${OUTPUT_DIR}/backups.txt
      ^-----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
touch "${OUTPUT_DIR}"/backups.txt


In smoke-test.sh line 35:
echo "Logs directory created at "${LOGS_DIR}/smoke
                                 ^---------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
echo "Logs directory created at ""${LOGS_DIR}"/smoke


In smoke-test.sh line 38:
oc project ${TEST_NAMESPACE}
           ^---------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
oc project "${TEST_NAMESPACE}"


In smoke-test.sh line 42:
${PYTHON_VENV_DIR}/bin/pip install -q -r smoke/requirements.txt
^----------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
"${PYTHON_VENV_DIR}"/bin/pip install -q -r smoke/requirements.txt


In smoke-test.sh line 44:
echo "Logs will be collected in "${TEST_SMOKE_OUTPUT_DIR}
                                 ^----------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
echo "Logs will be collected in ""${TEST_SMOKE_OUTPUT_DIR}"


In smoke-test.sh line 45:
${PYTHON_VENV_DIR}/bin/behave --junit --junit-directory ${TEST_SMOKE_OUTPUT_DIR} \
^----------------^ SC2086: Double quote to prevent globbing and word splitting.
                                                        ^----------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
"${PYTHON_VENV_DIR}"/bin/behave --junit --junit-directory "${TEST_SMOKE_OUTPUT_DIR}" \


In smoke-test.sh line 47:
                              smoke/features -D project_name=${TEST_NAMESPACE}                           
                                                             ^---------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
                              smoke/features -D project_name="${TEST_NAMESPACE}"                           


In smoke-test.sh line 48:
echo "Logs collected in "${TEST_SMOKE_OUTPUT_DIR}
                         ^----------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
echo "Logs collected in ""${TEST_SMOKE_OUTPUT_DIR}"

For more information:
  https://www.shellcheck.net/wiki/SC2155 -- Declare and assign separately to ...
  https://www.shellcheck.net/wiki/SC3010 -- In POSIX sh, [[ ]] is undefined.
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

File: smoke/README.md
Comment (gabemontero) [inline smoke/README.md:96]: Do you really want the sentence here bracketed by `<pre> </pre>`

It looks weird when a view this file from your PR in a browser, with markdown rendered.

Probably should remove the pre's

File: smoke/README.md
Comment (gabemontero) [inline smoke/README.md:56]:
```suggestion
</pre>

If the test case captured by the feature file is something which QA needs to run manually vs. a set of python based steps, then the contributor stops at the introduction of the feature file, and QA/Dev review for correctness.
```

File: smoke/README.md
Comment (gabemontero) [inline smoke/README.md:6]:
```suggestion
Hence, the reader should consider these tests the 'next phase' of testing after the [golang based tests](https://github.com/openshift/csi-driver-shared-resource/tree/master/test/e2e) that run in OpenShift CI prow jobs as part of merging pull requests with code changes.

Both QA and development have agreed to utilize this framework, and the description of test scenarios detailed below, as a replacement for direct communication between QA and development in OpenShift's internal polarion server.  QA will separately deal with taking the test scenarios descriptions here and importing into polarion for maintaining compatibility with the existing processes of the broader QA organization. 
```

**Review by gabemontero — CHANGES_REQUESTED** [review]
Just a few more minor items @jitendar-singh 

@rgarg1 @shwetha-h-p @adambkaplan FYI

File: .gitignore
Comment (adambkaplan) [inline .gitignore:3]: Is this something that our scripts/tooling generate? I would leave this out of a project .gitignore otherwise.

File: .gitignore
Comment (adambkaplan) [inline .gitignore:4]: Would it make sense to omit all `__pycache__` directories?
```suggestion
__pycache__/
```

File: hack/smoke-test.sh
Comment (adambkaplan) [reply hack/smoke-test.sh:7]: @gabemontero should we consider adding shellcheck to our `verify` test, perhaps as a follow-up PR?

File: smoke/README.md
Comment (adambkaplan) [reply smoke/README.md:6]: Personally, I would omit the suggested second paragraph. This references processes that are more internal to Red Hat and aren't germane to a customer or end user. I propose the following:

```suggestion
Hence, the reader should consider these tests the 'next phase' of testing after the [golang based tests](https://github.com/openshift/csi-driver-shared-resource/tree/master/test/e2e) that run in OpenShift CI prow jobs as part of merging pull requests with code changes. These provide another means of testing that are focused on the end user experience, in language and terms that end users can understand.
```

File: smoke/README.md
Comment (adambkaplan) [inline smoke/README.md:14]: Use triple-backticks for GitHub Markdown - ignore my slash since it breaks the suggestions
```suggestion
\```sh
$ oc login [cluster-server] -u username -p password --kubeconfig=kubeconfig
$ export KUBECONFIG=kubeconfig
$ make smoke
\```
```

File: smoke/README.md
Comment (adambkaplan) [inline smoke/README.md:27]:
```suggestion
\```
features/
features/everything.feature
features/steps/
features/steps/steps.py
\```
```

File: smoke/README.md
Comment (adambkaplan) [inline smoke/README.md:38]:
```suggestion
\```
features/
features/signup.feature
features/login.feature
features/account_details.feature
features/environment.py
features/steps/
features/steps/website.py
features/steps/utils.py
\```
```

File: smoke/README.md
Comment (adambkaplan) [inline smoke/README.md:56]:
```suggestion
\```
Feature: Fight or flight
  In order to increase the ninja survival rate,
  As a ninja commander
  I want my ninjas to decide whether to take on an
  opponent based on their skill levels

  Scenario: Weaker opponent
    Given the ninja has a third level black-belt
     When attacked by a samurai
     Then the ninja should engage the opponent

  Scenario: Stronger opponent
    Given the ninja has a third level black-belt
     When attacked by Chuck Norris
     Then the ninja should run for his life
\```
```

File: smoke/README.md
Comment (adambkaplan) [inline smoke/README.md:58]: Use `#` to create headers. `##` indicates an "h2" level header

```suggestion
## Python Step Implementations

Steps used in the scenarios are implemented in Python files in the “steps” directory. You can call these whatever you like as long as they use the python *.py file extension. You don’t need to tell behave which ones to use - it’ll use all of them.
```

File: smoke/README.md
Comment (adambkaplan) [reply smoke/README.md:96]: Hard agree here - use triple-backticks instead.

File: smoke/README.md
Comment (adambkaplan) [inline smoke/README.md:101]: Use a numbered list instead of html <p> tags
```suggestion
1. For each feature QE/Dev to create a feature file inside the smoke/features directory, the feature file name should be synonymus with the feature being developed.
2. Team to review the feature file and post approval QE to add the feature into polarion.
3. Once the feature file is approved then the contibutor can go ahead and work on the step definitions for the feature. The step definition file(.py) needs to be in the smoke/features/steps/ directory.
```

**Review by adambkaplan — CHANGES_REQUESTED** [review]
Leaving the code as is for the time being, in the interim please replace HTML tags with appropriate GitHub markdown. See https://docs.github.com/en/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

Also, commits should be squashed before merging.

File: smoke/README.md
Comment (gabemontero) [reply smoke/README.md:6]: If we are not going to document my suggestion on our team's evolving process here in this README, then I think I'll insist that it is documented *somewhere*

Perhaps https://source.redhat.com/groups/public/atomicopenshift/atomicopenshift_wiki/team_jenkinsbuild_api @adambkaplan ?  Or do you have another suggestion?

@rgarg1 @shwetha-h-p FYI

File: hack/smoke-test.sh
Comment (gabemontero) [reply hack/smoke-test.sh:7]: follow up PRs from anybody on the team would certainly be welcome for what you suggest @adambkaplan

Comment (jitendar-singh) [reply hack/smoke-test.sh]: Done

File: .gitignore
Comment (jitendar-singh) [reply .gitignore:3]: To run the smoke-test we need to export the KUBECONFIG. its needed. Its mentioned in README

File: smoke/README.md
Comment (rgarg1) [reply smoke/README.md]: Is there a team confluence page that exists? That might be an apt place to document internal information.

File: .gitignore
Comment (jitendar-singh) [reply .gitignore:4]: Not sure about that, but i can only see __pycache__ in smoke/features/steps

File: smoke/README.md
Comment (jitendar-singh) [reply smoke/README.md]: Done

File: hack/smoke-test.sh
Comment (jitendar-singh) [reply hack/smoke-test.sh]: Used the shellcheck & updated the script and see nothing now.

File: smoke/README.md
Comment (gabemontero) [reply smoke/README.md]: no we don't do confluence for this team that I'm aware of

the source wiki I noted above is what we have

we'll discuss in our team meeting most likely and make a decision there one way or the other

**Comment by gabemontero** [comment]
Aside from the getting consensus on the "where we fully document internal process details" question that sounds like an office hours item, I think I'm good with things here @jitendar-singh 

@adambkaplan - please indicate here if you are good with the changes to your requests outside of the "where we fully document internal process details" question

then once we get consensus on that process question, one of us will tag this for merge

File: smoke/features/steps/steps.py
Comment (prietyc123) [inline smoke/features/steps/steps.py:23]: @jitendar-singh Can we make it generic or comment it.

File: smoke/features/steps/steps.py
Comment (prietyc123) [inline smoke/features/steps/steps.py:24]: same as above 🙂

**Review by prietyc123 — CHANGES_REQUESTED** [review]

**Comment by gabemontero** [comment]
bump @jitendar-singh on @prietyc123 comments ... we can add the lgtm label once those are done.

@prietyc123 I assume you'll add the qe approved label when you are happy

also, I know @prietyc123 is in process wrt documenting our new sprint story qe proces on our team's wiki, so the exchanges with @adambkaplan above about minimizing the documentation of that process here in this PR's contents is sorted out from my point of view 

/assign @rolfedh 

for the docs approved label.  This PR is a internal QE process item and has no impact on docs.

any other further comments from you @adambkaplan ?

**Comment by adambkaplan** [comment]
> /docs-approved

/label docs-approved

**Comment by prietyc123** [comment]
@gabemontero @adambkaplan @jitendar-singh I have added the docs for the test flow here https://source.redhat.com/groups/public/atomicopenshift/atomicopenshift_wiki/team_jenkinsbuild_api

**Comment by gabemontero** [comment]
> @gabemontero @adambkaplan @jitendar-singh I have added the docs for the test flow here https://source.redhat.com/groups/public/atomicopenshift/atomicopenshift_wiki/team_jenkinsbuild_api

looks cool @prietyc123 

one note, per discussion in the google doc that was the drafting mechanism https://source.redhat.com/groups/public/atomicopenshift/atomicopenshift_wiki/automation__testing__build__jenkins - when this PR merges, let's add a link to this repo and the README @jitendar-singh in introducing here as part of the list of repos that follow this process.

I would also think what @jitendar-singh has also https://github.com/openshift/jenkins/tree/master/smoke would be part of that list.  Might also want to duplicate the README here in this PR over in that repo.

**Comment by gabemontero** [comment]
since @prietyc123 lgtm'ed this I'm assuming it is qe approved :-)

/label qe-approved

**Pull request merged**
